### ISSUE TITLE
Contract propagation bug?

### ISSUE THREAD

**Issue by panosru** [php-deal/framework#28]
I'm not sure if I'm doing something wrong but if I modify this:

https://github.com/php-deal/framework/blob/c508a19dd9b663dc195468854ed85effe151237f/demo/Demo/Account.php#L42-L51

into this:

```php
    /**
     * {@inheritdoc}
     */
    public function getBalance()
    {
        // return $this->balance;
        return 10;
    }
```

I should've receive the following error:
```
Fatal error: Uncaught DomainException: Invalid return value received from the assertion body, only boolean or void can be returned in deal/demo/Account.php on line 51

PhpDeal\Exception\ContractViolation: Contract $__result == $this->balance violated for Demo\Account->getBalance in deal/demo/Account.php on line 51
```

instead, I'm getting:
```
$ php demo.php
10
```

this:
```php
    /**
     * Returns current balance
     *
     * @Contract\Ensure("$__result == $this->balance")
     * @return float
     */
    public function getBalance()
    {
        // return $this->balance;
        return 10;
    }
```

returns the above-mentioned error as expected.

**Comment by panosru** [comment]
@icanhazstring I'm sorry for my late reply, I had a flight to catch, yes I expected the `@inheritdoc` to use the `@Contract\Ensure` from the parent.

from what I read [here](https://github.com/php-deal/framework#contract-propagation)

```php
class Foo extends FooParent
{
    /**
     * @param int $amount
     * @Contract\Verify("$amount != 1")
     * {@inheritdoc}
     */
    public function bar($amount)
    {
        ...
    }
}
    
class FooParent
{
    /**
     * @param int $amount
     * @Contract\Verify("$amount != 2")
     */
    public function bar($amount)
    {
        ...
    }
}
```
*Foo::bar does not accept '1' and '2' literals as a parameter.* 

That tells me that `Foo::bar` inherits the `@Contract\Verify` from `FooParent::bar` because of `@inheritdoc`, based on the example on the docs that should work:

```php
class Foo extends FooParent
{
    /**
     * @param int $amount
     * {@inheritdoc}
     */
    public function bar($amount)
    {
        ...
    }
}
    
class FooParent
{
    /**
     * @param int $amount
     * @Contract\Verify("$amount != 2")
     */
    public function bar($amount)
    {
        ...
    }
}
```

Which is basically but it seems that `@inheritdoc` works only when you have a `@Contract\*` in the method you add the `@inheritdoc`, otherwise is ignored.

The "problem" here is that `@inheritdoc` works only when you have specified in that docblock a `@Contract\*`, but `@inheritdoc` is widely used and having the library detecting `@inheritdoc` would bring more issues rather than solving the problem.

I suppose that `@inheritdoc` could work there where you use `@Contract\*` but if in my case you just want to inherit the parent contracts a separate annotation should be introduced `@Contract\Inherit` for instance.

**Comment by icanhazstring** [comment]
@panosru thanks for the explanation. Seems like a real bummer :wink:. Will try to look into that.

**Comment by icanhazstring** [comment]
Just cross referencing #30 with this, since I realized the exception you expect seems not correct.
This should be changed, and might be in the next major version.

Anyways, I dug into your problem and can confirm it. The problem with that is, the aspects are only working for methods with an annotation. I will try to figure out a good way to scan for inherited contracts.

**Comment by panosru** [comment]
@icanhazstring So, to visualise it, based on the solution you described, it would look like this:

```php
/**
 * @InheritContracts
 */
class Foo extends FooParent
{
    /**
     * @param int $amount
     * {@inheritdoc}
     */
    public function bar($amount)
    {
        ...
    }
}
    
class FooParent
{
    /**
     * @param int $amount
     * @Contract\Verify("$amount != 2")
     */
    public function bar($amount)
    {
        ...
    }
}
```

Tagging a class will then seek its methods for potential `@inheritdoc` in their dockblock, am I right?

**Comment by panosru** [comment]
@icanhazstring Thanks for the branch! I cloned it and I'm testing it but it seems to not working for me.

I have created a [debug repo](https://github.com/panosru/php-deal-debug) which you can clone and run `composer install` then `php test.php` will return `10` instead of error as it supposed to...

Also you had a type on your previous message, it should be `composer require php-deal/framework:dev-feature/inherit-contracts` *(with dash)*

**Comment by icanhazstring** [comment]
@panosru will give it a try and see what is wrong.

**Comment by icanhazstring** [comment]
@panosru as you can see I just provided a new PR solving your issue.
Feel free to look it up. Can you please try again if this implementation works for you.

If so I will merge this an tag a new release.

**Comment by icanhazstring** [comment]
@panosru just released a new version `0.5` with the new feature.

Closing this.
